tests/ubench/event_setup.cpp

```diff
@@ -0,0 +1,207 @@
+            events.push(e);
+        }
+
+        // pop from queue to form single sorted vector
+        std::vector<pev> staged_events;
+        staged_events.reserve(events.size());
+        while (auto e = events.pop_if_before(1.f)) {
+            staged_events.push_back(*e);
+        }
+        // sort the staged events in order of target id
+        std::stable_sort(
+            staged_events.begin(), staged_events.end(),
+            [](const pev& l, const pev& r) {return l.target.gid<r.target.gid;});
+
+        // TODO: calculate the partition ranges. This overhead is not included in
+        // this benchmark, however this method is that much slower already, that
+        // illustrating this wouldn't change the conclusions.
+
+        /*
+         // This is the "full" implementation that converts post_synaptic events
+         // to deliverable_events. Kepp it here in case the benchmark should
+         // be extended to test this step.
+        std::vector<deliverable_event> staged_events;
+        staged_events.reserve(events.size());
+        while (auto e = events.pop_if_before(2.f)) {
+            staged_events.emplace_back(
+                binner.bin(e->target.gid, e->time, 0.f),
+                // TODO: this should do a lookup in a hash table?
+                //get_target_handle(e->target),
+                target_handle(0, 0, e->target.gid),
+                e->weight);
+        }
+
+        binner.reset();
+        */
+
+        benchmark::ClobberMemory();
+    }
+}
+
+void n_queue(benchmark::State& state) {
+    using pev = postsynaptic_spike_event;
+    const std::size_t ncells = state.range(0);
+    const std::size_t ev_per_cell = state.range(1);
+
+    auto input_events = generate_inputs(ncells, ev_per_cell);
+
+    // state
+    std::vector<event_queue<pev>> event_lanes(ncells);
+    std::vector<size_t> part(ncells+1);
+
+    while (state.KeepRunning()) {
+        part[0] = 0;
+
+        // push events into the queue corresponding to target cell
+        for (const auto& e: input_events) {
+            event_lanes[e.target.gid].push(e);
+        }
+
+        // pop from queue to form single sorted vector
+        std::vector<pev> staged_events;
+        staged_events.reserve(input_events.size());
+        size_t i=0;
+        for (auto& lane: event_lanes) {
+            while (auto e = lane.pop_if_before(1.f)) {
+                staged_events.push_back(*e);
+            }
+            part[++i] = staged_events.size();
+        }
+
+        benchmark::ClobberMemory();
+    }
+}
+
+void n_vector(benchmark::State& state) {
+    using pev = postsynaptic_spike_event;
+    const std::size_t ncells = state.range(0);
+    const std::size_t ev_per_cell = state.range(1);
+
+    auto input_events = generate_inputs(ncells, ev_per_cell);
+
+    // state
+    std::vector<std::vector<pev>> event_lanes(ncells);
+    std::vector<size_t> part(ncells+1);
+    std::vector<decltype(input_events.begin())> ext(ncells);
```

Comment:
Logic is fine, but I reckon it would simplify the code and the reading thereof if `ext` just held a length/offset rather than the iterators.

Reply:
I made the change. Made some bits simpler, others more verbose.